### PR TITLE
[RFXCOM] Add new protocols

### LINKED ISSUE #2273: [RFXCOM] Add newer protocols
Add configuration options to enable the newer protocols decoded by rfxcom devices - specifically (but not limited to) FS20, homeconfort and keeloq.

### PR DESCRIPTION
**Pull request by mjagdis**
Support older firmware InterfaceMessage as per openhab1
Add config options for FS20, HomeConfort and KEELOQ
Add message handling for HomeConfort and KEELOQ

Closes #2273

Signed-off-by: Mike Jagdis <mjagdis@eris-associates.co.uk> (github: mjagdis)

### CODE REVIEW

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortMessage.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+        this.deviceId = Integer.parseInt(ids[0]);
+        houseCode = ids[1].charAt(0);
+        unitCode = Byte.parseByte(ids[2]);
+    }
+
+    @Override
+    public void convertFromState(RFXComValueSelector valueSelector, Type type) throws RFXComException {
+
+        switch (valueSelector) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortMessage.java]: I believe you convinced me that there was a one-on-one mapping between the `Type` and the `RFXComValueSelector` passed here, so maybe you can simplify the code to just the body of case.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java [hunk 2]
```diff
@@ -257,8 +301,27 @@ public void encodeMessage(byte[] data) throws RFXComException {
...
 
-            hardwareVersion1 = data[10];
-            hardwareVersion2 = data[11];
+            /**
+             * Different firmware versions have slightly different message formats.
+             * The firmware version numbering is unique to each hardware version
+             * but the location of the hardware version in the message is one of
+             * those things whose position varies. So we have to just look at the
+             * firmware version and pray. This condition below is taken from the
+             * openhab1-addons binding.
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java]: Great that you ported this.

Please credit the author in both the commit message,using also-by and in the file using the @author tag.

File: addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortTest.java [hunk 3]
```diff
@@ -8,20 +8,40 @@
...
+import org.openhab.binding.rfxcom.internal.messages.RFXComHomeConfortMessage;
+import org.openhab.binding.rfxcom.internal.messages.RFXComHomeConfortMessage.SubType;
+import org.openhab.binding.rfxcom.internal.messages.RFXComHomeConfortMessage.Commands;
+
 
 /**
  * Test for RFXCom-binding
  *
  * @author Martin van Wingerden
+ * @author Mike Jagdis
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortTest.java]: Maybe change mine to initial contribution of empty test and add after your name that you added the real test

File: addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComSecurity2MessageTest.java [hunk 4]
```diff
@@ -8,20 +8,38 @@
...
 
 import org.junit.Test;
-import org.openhab.binding.rfxcom.internal.exceptions.RFXComMessageNotImplementedException;
+import org.openhab.binding.rfxcom.internal.exceptions.RFXComException;
 
 /**
  * Test for RFXCom-binding
  *
  * @author Martin van Wingerden
+ * @author Mike Jagdis
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComSecurity2MessageTest.java]: Maybe change mine to initial contribution of empty test and add after your name that you added the real test

**Comment by martinvw** [comment]
Forcing a rebuild

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortMessage.java (hunk 1 above)
Comment (mjagdis) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortMessage.java]: I didn't want to end up with two different but equivalent ways of doing things. The whole RFXComValueSelector thing should be killed across the board at once.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java (hunk 2 above)
Comment (mjagdis) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java]: Done.

File: addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortTest.java (hunk 3 above)
Comment (mjagdis) [inline addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComHomeConfortTest.java]: Done.

File: addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComSecurity2MessageTest.java (hunk 4 above)
Comment (mjagdis) [inline addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComSecurity2MessageTest.java]: Done.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java (hunk 2 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessage.java]: @mjagdis can it be that we also have to port the actual updated handling of the firmware version, I just noted this line in a test:

https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.rfxcom.test/src/test/java/org/openhab/binding/rfxcom/internal/messages/RFXComInterfaceMessageTest.java#L62

**Pull request merged**